src/lib.rs

```diff
@@ -621,182 +654,203 @@ where
+                .chain_err(|| "invalid subprogram specification origin")?
+            {
+                return Ok(Some(name));
+            }
         }
 
         Ok(None)
     }
 
-    fn get_entry<'a>(
-        entry: &gimli::DebuggingInformationEntry<'a, 'a, gimli::EndianBuf<'input, Endian>>,
-        header: &'a gimli::CompilationUnitHeader<gimli::EndianBuf<'input, Endian>>,
-        abbrev: &'a gimli::Abbreviations,
-        attr: gimli::DwAt,
-    ) -> Result<Option<gimli::DebuggingInformationEntry<'a, 'a, gimli::EndianBuf<'input, Endian>>>> {
-        if let Some(gimli::AttributeValue::UnitRef(offset)) = entry
-            .attr_value(attr)
-            .map_err(|e| Error::from(ErrorKind::Gimli(e)))?
-        {
-            let mut entries = header.entries_at_offset(abbrev, offset)?;
-            let (_, entry) = entries.next_dfs()?.ok_or_else(|| {
-                ErrorKind::InvalidDebugSymbols(DebugInfoError::DanglingEntryOffset)
-            })?;
-            return Ok(Some(entry.clone()));
-        }
+    fn resolve_indirect_name(
+        &self,
+        attr: &gimli::AttributeValue<gimli::EndianBuf<'input, Endian>>,
+    ) -> Result<Option<gimli::EndianBuf<'input, Endian>>> {
+        let mut entries = match attr {
+            &gimli::AttributeValue::UnitRef(offset) => {
+                self.header.entries_at_offset(&self.abbrev, offset)?
+            }
+            // FIXME: handle AttributeValue::DebugInfoRef
+            _ => return Ok(None),
+        };
+        let (_, entry) = entries.next_dfs()?.ok_or_else(|| {
+            ErrorKind::InvalidDebugSymbols(DebugInfoError::DanglingEntryOffset)
+        })?;
 
-        Ok(None)
-    }
+        let mut name = None;
+        let mut abstract_origin = None;
+        let mut specification = None;
 
-    fn get_ranges(
-        entry: &gimli::DebuggingInformationEntry<gimli::EndianBuf<Endian>>,
-        debug_ranges: &gimli::DebugRanges<gimli::EndianBuf<Endian>>,
-        address_size: u8,
-        base_address: u64,
-    ) -> Result<Vec<gimli::Range>> {
-        if let Some(range) =
-            Self::parse_noncontiguous_ranges(entry, debug_ranges, address_size, base_address)?
+        let mut attrs = entry.attrs();
+        while let Some(attr) = attrs
+            .next()
+            .map_err(|e| Error::from(ErrorKind::Gimli(e)))
+            .chain_err(|| "invalid subprogram attribute")?
         {
-            return Ok(range);
-        }
-        if let Some(range) = Self::parse_contiguous_range(entry)?.map(|range| vec![range]) {
-            return Ok(range);
+            match attr.name() {
+                // For naming, we prefer the linked name, if available
+                gimli::DW_AT_linkage_name | gimli::DW_AT_MIPS_linkage_name => {
+                    if let Some(name) = attr.string_value(&self.debug_str) {
+                        return Ok(Some(name));
+                    }
+                }
+                gimli::DW_AT_name => {
```

Comment:
this is missing `if name.is_none()`

Reply:
ah, I guess you return early in the arm above, so it doesn't really matter..
should the function return if `name.is_some()` instead of calling `self.resolve_name` recursively below?

Reply:
Maybe. I'll see if I can make this a bit cleaner.